dpcpp/solver/batch_bicgstab_kernels.dp.cpp

```diff
@@ -33,39 +33,247 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+                    ValueType* const x_global_entry =
+                        gko::batch::multi_vector::batch_item_ptr(
+                            x_values, 1, num_rows, batch_id);
+                    apply_kernel<StopType, n_shared_total, sg_kernel_all>(
+                        sconf, max_iters, res_tol, logger, prec,
+                        mat_global_entry, b_global_entry, x_global_entry,
+                        num_rows, mat.get_single_item_num_nnz(),
+                        static_cast<ValueType*>(slm_values.get_pointer()),
+                        item_ct1, workspace);
+                });
+        });
+    }
+
+    template <typename BatchMatrixType, typename PrecType, typename StopType,
+              typename LogType>
+    void call_kernel(
+        LogType logger, const BatchMatrixType& mat, PrecType prec,
+        const gko::batch::multi_vector::uniform_batch<const ValueType>& b,
+        const gko::batch::multi_vector::uniform_batch<ValueType>& x) const
+    {
+        using real_type = gko::remove_complex<ValueType>;
+        const size_type num_batch_items = mat.num_batch_items;
+        const auto num_rows = mat.num_rows;
+        const auto num_rhs = b.num_rhs;
+        GKO_ASSERT(num_rhs == 1);
+
+        auto device = exec_->get_queue()->get_device();
+        auto group_size =
+            device.get_info<sycl::info::device::max_work_group_size>();
+        if (group_size > num_rows) group_size = get_group_size(num_rows);
+
+        size_type shmem_per_blk =
+            device.get_info<sycl::info::device::local_mem_size>() -
+            (group_size + 5) * sizeof(ValueType) -
+            2 * sizeof(
+                    real_type);  // reserve 5 for intermediate rho-s, norms,
+                                 // alpha, omega, temp and for reduce_over_group
+        if (shmem_per_blk < 0) shmem_per_blk = 0;
+        const int shared_gap = num_rows;
+        const size_type prec_size = PrecType::dynamic_work_size(
+            shared_gap, mat.get_single_item_num_nnz());
+        const auto sconf =
+            gko::kernels::batch_bicgstab::compute_shared_storage<PrecType,
+                                                                 ValueType>(
+                shmem_per_blk, shared_gap, mat.get_single_item_num_nnz(),
+                b.num_rhs);
+        const size_t shared_size =
+            sconf.n_shared * shared_gap + (sconf.prec_shared ? prec_size : 0);
+        auto workspace = gko::array<ValueType>(
+            exec_,
+            sconf.gmem_stride_bytes * num_batch_items / sizeof(ValueType));
+        assert(sconf.gmem_stride_bytes % sizeof(ValueType) == 0);
+
+        ValueType* const workspace_data = workspace.get_data();
+        int n_shared_total = sconf.n_shared + int(sconf.prec_shared);
+
+        // template
+        // launch_apply_kernel<StopType, SIMDLEN, n_shared_total, sg_kernel_all>
+        if (num_rows <= 32 && n_shared_total == 10)
```

Comment:
cuda/hip uses 9 vectors in shmem. Why does this check for 10? Also the kernel only checks until `n_shared_total == 9`

Reply:
the strategy is slightly different. Here the count includes the prec_shared vector.  The number of shared vectors is always 9, so you can only check until 9. If it is greater than 9, then you know that the prec is also in shared memory.

Reply:
but isn't that what `storage_config::prec_shared` is there for?

Reply:
I think it is a bit easier with looking at `n_shared` as 10 vectors. Otherwise, `prec_shared` will need to be a template parameter as well. But I understand your point that it makes the cuda/dpcpp kernels more confusing to compare.

Reply:
I would prefer the additional template parameter then. But that might also be done later.